pgdate: parse infinity/-infinity as max/min supported timestamp

otan opened the pull request:
Refs: #41564

Unsure of whether to backport this as it makes queries like #50301 unusable.

----

This is a stop gap measure to support people evaluating to a super high
or super low timestamp, e.g. `CREATE USER ... VALID UNTIL 'infinity'`.
As we don't support infinity proper yet, make `infinity`/`-infinity` as
a timestamp evaluate to the highest/lowest supported value.

Release note (sql change): Using `infinity` previously
evaluated to a negative, i.e. "-292277022365-05-08T08:17:07Z".
This has been fixed to be the maximum supported timestamp in Postgres
that is not infinity. Likewise, `-infinity` is the smallest supported
value. Note this does currently does not behave exactly like `infinity`
in Postgres (this is a work in progress and may be resolved later).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50311)
<!-- Reviewable:end -->

otan committed:
pgdate: parse infinity/-infinity as max/min supported timestamp

This is a stop gap measure to support people evaluating to a super high
or super low timestamp, e.g. `CREATE USER ... VALID UNTIL 'infinity'`.
As we don't support infinity proper yet, make `infinity`/`-infinity` as
a timestamp evaluate to the highest/lowest supported value.

Release note (sql change): Using `infinity` previously
evaluated to a negative, i.e. "-292277022365-05-08T08:17:07Z".
This has been fixed to be the maximum supported timestamp in Postgres
that is not infinity. Likewise, `-infinity` is the smallest supported
value. Note this does currently does not behave exactly like `infinity`
in Postgres (this is a work in progress and may be resolved later).

RichardJCai commented:
With this stop gap, I wonder how we'll handle moving to actually supporting `infinity` and `-infinity` as a timestamp in the future. It might be hard to migrate in the future. 

Also I'm just curious, what are we missing to actually support `infinity` as a timestamp?

otan commented:
> With this stop gap, I wonder how we'll handle moving to actually supporting infinity and -infinity as a timestamp in the future. It might be hard to migrate in the future.

future generations of CRDB can store "infinity" as a special value. it won't backup/restore the same, but we have that problem already.

> Also I'm just curious, what are we missing to actually support infinity as a timestamp?

we just need to plumb a lot of if statements everywhere. i don't have time to work on that atm (and tbh, it's probably a good first issue)

RichardJCai commented:
> > With this stop gap, I wonder how we'll handle moving to actually supporting infinity and -infinity as a timestamp in the future. It might be hard to migrate in the future.
> 
> future generations of CRDB can store "infinity" as a special value - but for now it'll be stored as a super high / super low value. it won't backup/restore the same, but we have that problem already in the current setup.
> 
> > Also I'm just curious, what are we missing to actually support infinity as a timestamp?
> 
> we just need to plumb a lot of if statements everywhere and auditing of datetime related builtins. i don't have time to work on that atm (and tbh, it's probably a good first issue)

That's true this stop gap doesn't introduce any new problems (I think) in the meanwhile and improves some functionality. Seems reasonable to me I think, as long as we note somewhere that this is the behaviour until we support infinity.

rohany commented on pkg/util/timeutil/pgdate/parsing.go:
```diff
@@ -65,9 +64,21 @@ var (
+	// We should probably use the next microsecond after this value, i.e. timeutil.Unix(9224318016000, 0).
+	// Postgres uses math.MaxInt64 microseconds as the infinity value.
+	// See: https://github.com/postgres/postgres/blob/42aa1f0ab321fd43cbfdd875dd9e13940b485900/src/include/datatype/timestamp.h#L107.
+	TimeInfinity = timeutil.Unix(9224318016000-1, 999999000)
```
Just to double check, we were already returning these when seeing "infinity", but they just overflowed here?

otan replied:
we were returning a value that overflowed to negative date value

otan commented:
bors r=rohany

tftr!

RichardJCai commented:
> Refs: #41564
> 
> Unsure of whether to backport this as it makes queries like #50301 unusable.
> 
> This is a stop gap measure to support people evaluating to a super high
> or super low timestamp, e.g. `CREATE USER ... VALID UNTIL 'infinity'`.
> As we don't support infinity proper yet, make `infinity`/`-infinity` as
> a timestamp evaluate to the highest/lowest supported value.
> 
> Release note (sql change): Using `infinity` previously
> evaluated to a negative, i.e. "-292277022365-05-08T08:17:07Z".
> This has been fixed to be the maximum supported timestamp in Postgres
> that is not infinity. Likewise, `-infinity` is the smallest supported
> value. Note this does currently does not behave exactly like `infinity`
> in Postgres (this is a work in progress and may be resolved later).

Wait why does backporting this make the query in https://github.com/cockroachdb/cockroach/issues/50301 unusable? Isn't it closer to the behaviour we actually want?

otan commented:
oops, i mean it changes behaviour.

RichardJCai commented:
> oops, i mean it changes behaviour.

But it does make the behaviour strictly better right?
